scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchGenerator.java

```diff
@@ -0,0 +1,536 @@
+              valueFilterExpression = new AttributeComparisonExpression(new AttributeReference("type"), CompareOperator.EQ, type);
+            } else if (parseData.originalObject instanceof String || parseData.originalObject instanceof Number) {
+              String toString = parseData.originalObject.toString();
+              valueFilterExpression = new AttributeComparisonExpression(new AttributeReference("value"), CompareOperator.EQ, toString);
+            } else if(parseData.originalObject instanceof Enum) {
+              Enum<?> tempEnum = (Enum<?>)parseData.originalObject;
+              valueFilterExpression = new AttributeComparisonExpression(new AttributeReference("value"), CompareOperator.EQ, tempEnum.name());
+            } else {
+              if (parseData.originalObject != null) {
+                log.debug("Attribute: {} doesn't implement TypedAttribute, can't create ValueFilterExpression", parseData.originalObject.getClass());
+              } else {
+                log.debug("Attribute: null doesn't implement TypedAttribute, can't create ValueFilterExpression");
+              }
+              valueFilterExpression = new AttributeComparisonExpression(new AttributeReference("value"), CompareOperator.EQ, "?");
+            }
+            processingMultiValued = false;
+            processedMultiValued = true;
+          }
+        } else {
+          Schema.Attribute attribute = parseData.ac.getAttribute(pathPart);
+
+          if (attribute != null) {
+            if (processedMultiValued) {
+              subAttributes.add(pathPart);
+            } else {
+              log.debug("Adding {} to attributeReferenceList", pathPart);
+              attributeReferenceList.add(pathPart);
+            }
+
+            parseData.traverseObjects(pathPart, attribute);
+
+            if (patchOpType == PatchOperation.Type.REPLACE &&
+              (parseData.resourceObject != null && parseData.resourceObject instanceof Collection && !((Collection<?>)parseData.resourceObject).isEmpty()) &&
+              (parseData.originalObject == null ||
+                (parseData.originalObject instanceof Collection && ((Collection<?>)parseData.originalObject).isEmpty()))) {
+              patchOpType = PatchOperation.Type.ADD;
+            }
+
+            if (attribute.isMultiValued()) {
+              processingMultiValued = true;
+            } else if (attribute.getType() != Schema.Attribute.Type.COMPLEX) {
+              done = true;
+            }
+          }
+        }
+        ++i;
+      }
+
+      if (patchOpType == PatchOperation.Type.REPLACE && (parseData.resourceObject == null ||
+        (parseData.resourceObject instanceof Collection && ((Collection<?>)parseData.resourceObject).isEmpty()))) {
+        patchOpType = PatchOperation.Type.REMOVE;
+        valueNode = null;
+      }
+
+      if (patchOpType == PatchOperation.Type.REPLACE && parseData.originalObject == null) {
+        patchOpType = PatchOperation.Type.ADD;
+      }
```

Comment:
I see this is part what I mentioned in another comment - a replace operation actually being an add/remove depending on the original or target resource having `null`